File: 41-first-missing-positive/41-first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();    
        for( int i = 0; i < n; i++ )  {
            while((nums[i] > 0) &&
                  (nums[i] - 1 < n ) &&
                  (nums[i] != i + 1) &&
                  (nums[i] != nums[nums[i] - 1] ) ) {
                int d = nums[i] - 1;
                swap(nums[i], nums[d]); 
            }
        }
        
        for( int i = 0; i < n; i++) {
            if( nums[i] != i + 1) {
                return i + 1;
            }
        }
        
        return n + 1;
    }
};

File: 41-first-missing-positive/41-first-missing-positive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        for i in range(len(nums)):
            while nums[i] != i + 1:
                d = nums[i] - 1
                if 0 <= d < len(nums) and nums[i] != nums[d]:
                    nums[i], nums[d] = nums[d], nums[i]
                else:
                    break;
                    
        for i in range(len(nums)):
            if nums[i] != i + 1:
                return i + 1
        
        return len(nums)+1
            
        